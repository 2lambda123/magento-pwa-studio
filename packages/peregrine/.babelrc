{
  "plugins": [
    "transform-object-rest-spread",
    "transform-class-properties",
    ["transform-react-jsx", { "pragma": "createElement" }],
    ["transform-runtime", {
      "helpers": true,
      "polyfill": false, // polyfills will be handled by preset-env
      "regenerator": false
    }]
  ],
  "presets": [
    ["env", {
      "targets": {
        "browsers": ["> 5%"]
      },
      "modules": false
    }]
  ],
  "env": {
    "test": {
      "plugins": [
        "transform-es2015-modules-commonjs",
        "transform-object-rest-spread",
        "transform-class-properties",
        ["transform-react-jsx", { "pragma": "createElement" }]
      ]
    }
  }
}
