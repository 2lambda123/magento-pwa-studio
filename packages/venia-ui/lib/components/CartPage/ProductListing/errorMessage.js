import React from 'react';
import { FormattedMessage } from 'react-intl';
import { shape, string } from 'prop-types';
import { useStyle } from '../../../classify';
import defaultClasses from './errorMessage.css';

const ErrorMessage = props => {
    const { error } = props;
    const classes = useStyle(defaultClasses, props.classes);

    if (!error || !error.message) {
        return null;
    }

    const errorText = error.message.includes('qty') ? (
        <FormattedMessage
            id="stockStatusMessage.message"
            defaultMessage="An item in your cart is currently out-of-stock and must be removed in order to Checkout."
        />
    ) : (
        error.message
    );

    return <div className={classes.root}>{errorText}</div>;
};

ErrorMessage.defaultProps = {
    classes: null,
    error: null
};

ErrorMessage.propTypes = {
    classes: shape({
        root: string
    }),
    error: shape({
        message: string
    })
};

export default ErrorMessage;
