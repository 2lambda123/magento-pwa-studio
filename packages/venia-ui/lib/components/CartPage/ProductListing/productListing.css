.root {
    display: grid;
    grid-gap: 2rem 1rem;
}

.rootMasked {
    composes: root;
    opacity: 0.5;
    pointer-events: none;
}
