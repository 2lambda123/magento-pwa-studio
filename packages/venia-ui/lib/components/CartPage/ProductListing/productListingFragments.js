import gql from 'graphql-tag';

export const ProductListingFragment = gql`
    fragment ProductListingFragment on Cart {
        id
        items {
            id
            product {
                id
                name
                small_image {
                    url
                }
            }
            prices {
                price {
                    currency
                    value
                }
            }
            quantity
            ... on ConfigurableCartItem {
                configurable_options {
                    id
                    option_label
                    value_label
                }
            }
        }
    }
`;
