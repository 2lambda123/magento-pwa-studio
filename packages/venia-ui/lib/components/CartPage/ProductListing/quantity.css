.root {
    text-align: center;
    display: grid;
    grid-template-columns: 1fr 4rem 1fr;
    grid-column-gap: 0.5rem;
    align-items: center;
    justify-items: center;
}

.wrap {
    display: flex;
    align-items: center;
    justify-content: center;
}

.label:not(:focus):not(:active) {
    position: absolute;
    overflow: hidden;
    width: 1px;
    height: 1px;
    clip: rect(0 0 0 0);
    clip-path: inset(100%);
    white-space: nowrap;
}

.button {
    align-items: center;
    background-color: white;
    border: 1px solid rgb(var(--venia-global-color-gray-darker));
    border-radius: 50%;
    display: inline-flex;
    height: 2.25rem;
    justify-content: center;
    width: 2.25rem;
}

.button:disabled {
    border-color: rgb(var(--venia-global-color-gray-darker)) !important;
    background-color: #d0d0d0 !important;
    opacity: 0.5;
    cursor: not-allowed;
}

.button:focus {
    border-color: rgb(var(--venia-global-color-teal));
    outline: none;
}

.button:hover {
    border-color: rgb(var(--venia-global-color-teal));
}

.button:active {
    border-color: rgb(var(--venia-global-color-teal));
}

.button_decrement {
    composes: button;
}

.button_increment {
    composes: button;
}

.icon {
    composes: root from '../../Icon/icon.css';
}

.input {
    composes: input from '../../TextInput/textInput.css';
    text-align: center;
}
