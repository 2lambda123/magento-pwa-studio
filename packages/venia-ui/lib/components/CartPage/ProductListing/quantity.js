import React from 'react';
import { func, number, string } from 'prop-types';

import { Minus as MinusIcon, Plus as PlusIcon } from 'react-feather';
import { Form } from 'informed';
import { useQuantity } from '@magento/peregrine/lib/talons/CartPage/ProductListing/useQuantity';

import Button from '../../Button';
import defaultClasses from './quantity.css';
import Icon from '../../Icon';
import TextInput from '../../TextInput';
import { mergeClasses } from '../../../classify';

const QuantityFields = props => {
    const classes = mergeClasses(defaultClasses, props.classes);
    const { initialValue, itemId, label, min, onChange } = props;

    const talonProps = useQuantity({
        initialValue,
        min,
        onChange
    });

    const {
        isDecrementDisabled,
        isIncrementDisabled,
        handleBlur,
        handleDecrement,
        handleIncrement,
        maskInput
    } = talonProps;

    return (
        <div className={classes.root}>
            <label className={classes.label} htmlFor={itemId}>
                {label}
            </label>
            <Button
                aria-label={'Decrease Quantity'}
                priority="normal"
                className={classes.button_decrement}
                disabled={isDecrementDisabled}
                onClick={handleDecrement}
                type="button"
            >
                <Icon className={classes.icon} src={MinusIcon} size={22} />
            </Button>
            <TextInput
                aria-label="Item Quantity"
                field="quantity"
                id={itemId}
                mask={maskInput}
                onBlur={handleBlur}
                min={min}
                type="number"
            />
            <Button
                aria-label={'Increase Quantity'}
                className={classes.button_increment}
                priority="normal"
                disabled={isIncrementDisabled}
                onClick={handleIncrement}
                type="button"
            >
                <Icon className={classes.icon} src={PlusIcon} size={22} />
            </Button>
        </div>
    );
};

const Quantity = props => {
    return (
        <Form
            initialValues={{
                quantity: props.initialValue
            }}
        >
            <QuantityFields {...props} />
        </Form>
    );
};

Quantity.propTypes = {
    initialValue: number,
    itemId: string,
    label: string,
    min: number,
    onChange: func
};

Quantity.defaultProps = {
    label: 'Quantity',
    min: 0,
    initialValue: 1,
    onChange: () => {}
};

export default Quantity;
