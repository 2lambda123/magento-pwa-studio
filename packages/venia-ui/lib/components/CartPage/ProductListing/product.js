import React from 'react';
import gql from 'graphql-tag';
import { useProduct } from '@magento/peregrine/lib/talons/CartPage/ProductListing/useProduct';
import { Price } from '@magento/peregrine';

import { mergeClasses } from '../../../classify';
import Kebab from '../../MiniCart/kebab';
import ProductOptions from '../../MiniCart/productOptions';
import Section from '../../MiniCart/section';
import Image from '../../Image';
import defaultClasses from './product.css';
import { GET_PRODUCT_LISTING } from './productListing';
import { PriceSummaryQuery } from '../PriceSummary/priceSummary';

const IMAGE_SIZE = 100;

const Product = props => {
    const { item } = props;
    const talonProps = useProduct({
        item,
        refetchPriceQuery: PriceSummaryQuery,
        refetchCartQuery: GET_PRODUCT_LISTING,
        removeItemMutation: REMOVE_ITEM_MUTATION
    });
    const {
        handleEditItem,
        handleRemoveFromCart,
        handleToggleFavorites,
        isFavorite,
        isRemoving,
        product
    } = talonProps;
    const { currency, image, name, options, quantity, unitPrice } = product;

    const classes = mergeClasses(defaultClasses, props.classes);
    const rowMask = isRemoving ? classes.mask : '';

    return (
        <li className={`${classes.root} ${rowMask}`}>
            <Image
                alt={name}
                classes={{ image: classes.image, root: classes.imageContainer }}
                width={IMAGE_SIZE}
                resource={image}
            />
            <div className={classes.details}>
                <span className={classes.name}>{name}</span>
                <ProductOptions
                    options={options}
                    classes={{
                        options: classes.options,
                        optionLabel: classes.optionLabel
                    }}
                />
                <span className={classes.price}>
                    <Price currencyCode={currency} value={unitPrice} />
                    {' ea.'}
                </span>
                {/** Quantity Selection to be completed by PWA-119. */}
                <div className={classes.quantity}>- {quantity} +</div>
            </div>
            <Kebab classes={{ root: classes.kebab }} disabled={true}>
                <Section
                    text={
                        isFavorite
                            ? 'Remove from favorites'
                            : 'Move to favorites'
                    }
                    onClick={handleToggleFavorites}
                    icon="Heart"
                    isFilled={isFavorite}
                    classes={{ text: classes.sectionText }}
                />
                <Section
                    text="Edit item"
                    onClick={handleEditItem}
                    icon="Edit2"
                    classes={{ text: classes.sectionText }}
                />
                <Section
                    text="Remove from cart"
                    onClick={handleRemoveFromCart}
                    icon="Trash"
                    classes={{ text: classes.sectionText }}
                />
            </Kebab>
        </li>
    );
};

export default Product;

export const REMOVE_ITEM_MUTATION = gql`
    mutation removeItem($cartId: String!, $itemId: Int!) {
        removeItemFromCart(input: { cart_id: $cartId, cart_item_id: $itemId }) {
            cart {
                id
            }
        }
    }
`;
