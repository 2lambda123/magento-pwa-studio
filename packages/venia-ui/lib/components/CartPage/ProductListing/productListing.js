import React, { Fragment } from 'react';
import gql from 'graphql-tag';
import { useProductListing } from '@magento/peregrine/lib/talons/CartPage/ProductListing/useProductListing';

import { mergeClasses } from '../../../classify';
import LoadingIndicator from '../../LoadingIndicator';
import EditModal from './EditModal';
import defaultClasses from './productListing.css';
import Product from './product';
import { ProductListingFragment } from './productListingFragments';

const ProductListing = props => {
    const { setIsCartUpdating } = props;
    const talonProps = useProductListing({
        queries: {
            getProductListing: GET_PRODUCT_LISTING
        }
    });
    const { activeEditItem, isLoading, items, setActiveEditItem } = talonProps;

    const classes = mergeClasses(defaultClasses, props.classes);

    if (isLoading) {
        return <LoadingIndicator>{`Fetching Cart...`}</LoadingIndicator>;
    }

    if (items.length) {
        const productComponents = items.map(product => (
            <Product
                item={product}
                key={product.id}
                setActiveEditItem={setActiveEditItem}
                setIsCartUpdating={setIsCartUpdating}
            />
        ));

        return (
            <Fragment>
                <ul className={classes.root}>{productComponents}</ul>
                <EditModal
                    item={activeEditItem}
                    setIsCartUpdating={setIsCartUpdating}
                />
            </Fragment>
        );
    }

    return null;
};

export const GET_PRODUCT_LISTING = gql`
    query getProductListing($cartId: String!) {
        cart(cart_id: $cartId) @connection(key: "Cart") {
            id
            ...ProductListingFragment
        }
    }
    ${ProductListingFragment}
`;

export default ProductListing;
