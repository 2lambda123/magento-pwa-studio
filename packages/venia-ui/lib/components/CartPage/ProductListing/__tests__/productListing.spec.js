import React from 'react';
import { useLazyQuery } from '@apollo/react-hooks';
import { createTestInstance } from '@magento/peregrine';

import LoadingIndicator from '../../../LoadingIndicator';
import ProductListing from '../productListing';

jest.mock('../../../../classify');
jest.mock('@apollo/react-hooks', () => {
    return { useLazyQuery: jest.fn() };
});

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const state = { cartId: 'cart123' };
    const api = {};
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});

jest.mock('../product', () => 'Product');

test('renders loading indicator while data fetching', () => {
    useLazyQuery.mockReturnValueOnce([
        () => {},
        {
            loading: true
        }
    ]);

    const { root } = createTestInstance(<ProductListing />);
    expect(root.findByType(LoadingIndicator)).toBeDefined();
});

test('renders string with no items in cart', () => {
    useLazyQuery.mockReturnValueOnce([
        () => {},
        {
            called: true,
            loading: false,
            error: null,
            data: {
                cart: {
                    items: []
                }
            }
        }
    ]);
    const tree = createTestInstance(<ProductListing />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders list of products with items in cart', () => {
    useLazyQuery.mockReturnValueOnce([
        () => {},
        {
            called: true,
            loading: false,
            error: null,
            data: {
                cart: {
                    items: ['1', '2', '3']
                }
            }
        }
    ]);
    const tree = createTestInstance(<ProductListing />);

    expect(tree.toJSON()).toMatchSnapshot();
});
