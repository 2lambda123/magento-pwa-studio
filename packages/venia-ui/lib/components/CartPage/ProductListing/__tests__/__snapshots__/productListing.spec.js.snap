// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders list of products with items in cart 1`] = `
<ul
  className="root"
>
  <Product
    item="1"
  />
  <Product
    item="2"
  />
  <Product
    item="3"
  />
</ul>
`;

exports[`renders string with no items in cart 1`] = `
<h3>
  There are no items in your cart.
</h3>
`;
